File: 763. Partition Labels/solution.py
class Solution:
    def partitionLabels(self, S: str) -> List[int]:
        sol = []
        n = len(S)
        
        # edge case
        if not n:
            return sol
        
        # gather all intervals for every lowercase letter, O(n)
        pos = [None]*26
        for i in range(n):
            index = ord(S[i]) - ord('a')
            if pos[index]:
                pos[index][1] = i
            else:
                pos[index] = [i, i]
                
        # convert to intervals array, O(1)
        intervals = []
        for p in pos:
            if p:
                intervals.append(p)
                
        # sort intervals and merge overlapping intervals, O(1)
        intervals.sort(key = lambda x: (x[0], x[1]))
        merged = [intervals[0]]
        for i in intervals[1:]:
            if i[0] < merged[-1][1]:
                merged[-1][1] = max(merged[-1][1], i[1])
            else:
                merged.append(i)
                
        # get interval lengths, O(1)
        for i in merged:
            sol.append(i[1] - i[0] + 1)
            
        return sol
            

File: 763. Partition Labels/solution.java
class Solution {
    public List<Integer> partitionLabels(String S) {
        int[] starts = new int[26];
        int[] ends = new int[26];
        
        // get start and end position
        for (int i = 0; i < 26; i++) {
            starts[i] = -1;
            ends[i] = -1;
        }
        
        int pos;
        for (int i = 0; i < S.length(); i++) {
            pos = Character.getNumericValue(S.charAt(i))-Character.getNumericValue('a');
            if (starts[pos] == -1) {
                starts[pos] = i;
            }
            ends[pos] = i;
        }
        
        // convert to merge interval problem
        List<int[]> intervals = new ArrayList<int[]>();
        for(int i = 0; i < 26; i++) {
            if (starts[i] != -1) {
                intervals.add(new int[]{starts[i], ends[i]});
            }
        }
        
        // merge intervals
        Collections.sort(intervals, (a, b) -> a[0] - b[0]);
        int start, end;
        List <Integer> sol = new ArrayList<Integer>();
        start = intervals.get(0)[0];
        end = intervals.get(0)[1];
        int i = 1;
        while (i < intervals.size()) {
            if (intervals.get(i)[0] <= end) {
                end = Math.max(intervals.get(i)[1], end);
            } else {
                sol.add(end - start + 1);
                start = intervals.get(i)[0];
                end = intervals.get(i)[1];
            }
            i++;
        }
        
        sol.add(end - start + 1);
        
        return sol;
    }
}
